Userland/Applications/PackageManager/InstalledPort.cpp

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2023, Liav A. <liavalb@hotmail.co.il>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Function.h>
+#include <Applications/PackageManager/InstalledPort.h>
+#include <LibCore/File.h>
+#include <LibCore/System.h>
+
+ErrorOr<HashMap<String, InstalledPort>> InstalledPort::read_ports_database()
+{
+    auto file = TRY(Core::File::open("/usr/Ports/installed.db"sv, Core::File::OpenMode::Read));
+    auto buffered_file = TRY(Core::InputBufferedFile::create(move(file)));
+    auto buffer = TRY(ByteBuffer::create_uninitialized(PAGE_SIZE));
+
+    HashMap<String, InstalledPort> ports;
+    while (TRY(buffered_file->can_read_line())) {
+        auto line = TRY(buffered_file->read_line(buffer));
+        if (line.starts_with("dependency"sv)) {
+            auto parts = line.split_view(' ');
+            auto type = InstalledPort::Type::Dependency;
+            // FIXME: Add versioning when printing these ports!
+            auto name = TRY(String::from_utf8(parts[1]));
+            TRY(ports.try_set(name, InstalledPort { TRY(String::from_utf8(parts[1])), type, TRY(String::from_utf8(""sv)) }));
+        } else {
+            auto parts = line.split_view(' ');
+            VERIFY(parts.size() == 3);
+            auto type = InstalledPort::Type::Auto;
+            if (parts[0] == "manual"sv) {
+                type = InstalledPort::Type::Manual;
+            } else if (parts[0] == "auto"sv) {
+                type = InstalledPort::Type::Auto;
+            } else {
+                VERIFY_NOT_REACHED();
+            }
+            auto name = TRY(String::from_utf8(parts[1]));
+            TRY(ports.try_set(name, InstalledPort { name, type, TRY(String::from_utf8(parts[2])) }));
+        }
+    }
+    return ports;
+}
+
+ErrorOr<void> InstalledPort::for_each_by_type(HashMap<String, InstalledPort>& ports_database, InstalledPort::Type type, Function<ErrorOr<void>(InstalledPort const&)> callback)
+{
+    VERIFY(!ports_database.is_empty());
+    VERIFY(type != InstalledPort::Type::Auto);
```

Comment:
These two `VERIFY`s seem pretty arbitrary - why not allow enumerating auto ports? Why can't we deal with an empty ports database?